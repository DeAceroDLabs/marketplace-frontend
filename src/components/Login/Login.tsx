import React, { useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import Section from "components/common/Section";
import View from "components/common/View";
import UserContext from "config/userContext";

const Login: React.FunctionComponent = () => {
  const navigate = useNavigate();
  const { username, setUser } = useContext(UserContext);
  const [user, setUsername] = useState(username);

  const submitUser = () => {
    setUser(user);
    navigate(`../`);
  };

  useEffect(() => {
    if (username) {
      navigate(`../`);
    }
  }, [username, navigate]);

  return (
    <View>
      <Section title="Login" variant="margin-title">
        <form onSubmit={submitUser}>
          <input
            value={user}
            type="text"
            placeholder="username"
            onChange={(e) => setUsername(e.currentTarget.value)}
          />
          <input type="text" placeholder="password" />
          <button type="submit">Log In</button>
        </form>
      </Section>
    </View>
  );
};

export default Login;
