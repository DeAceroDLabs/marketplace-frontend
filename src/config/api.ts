import { API, Category, Product, ZipCodeLocation } from "./api.types";
import axios from "axios";
import { USER_API } from "./user.api.types";

const fetchData = (url: string) => {
  async function fetchUrl() {
    const response = await fetch(url);
    const json = await response.json();
    return json;
  }
  return fetchUrl();
};

const postRequest = async (url: string, body: any) => {
  const response = await axios.post(USER_API.createUser, body);
  return response;
};

const fetchCategories = (): Promise<Category[]> => {
  return fetchData(API.categories);
};

const fetchProducts = (categoryId: number): Promise<Product[]> => {
  return fetchData(API.products(categoryId));
};

const getLocation = (zipCode: string): Promise<ZipCodeLocation> => {
  return fetchData(API.location(zipCode)).then((data) => {
    if (data["location"]) {
      const neighborhoodsObject = data["location"]["neighborhood"];
      const neighborhoods = Object.keys(neighborhoodsObject);
      const neighborhoodsArray = neighborhoods.map((neighborhood) => {
        return {
          value: neighborhood,
          label: neighborhoodsObject[neighborhood],
        };
      });
      data["location"]["neighborhood"] = neighborhoodsArray;
    }
    return data;
  });
};

const searchProducts = (query: string): Promise<any> => {
  const body = {
    q: query,
  };
  return postRequest(API.search, body);
};

export {
  postRequest,
  fetchCategories,
  fetchProducts,
  searchProducts,
  getLocation,
};
