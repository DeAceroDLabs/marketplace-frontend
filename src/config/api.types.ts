export const API = {
  categories: "categories",
  products: (categoryId: number) => `categories/${categoryId}/products`,
  search: "/search",
  location: (zipCode: string) => `locations/postalCode/${zipCode}`,
};

export type Category = {
  categoryId: number;
  categoryName: string;
  imageUrl: string;
};

export type Article = {
  articleId: string;
};

export type Product = {
  productId: number;
  productName: string;
  categoryName: string;
  categoryId: number;
  articles: Article[];
  categoryImageUrl: string;
};

export type ZipCodeLocation = {
  countryName: string;
  location: Location;
};

export type Location = {
  cityId: string;
  city: string;
  stateId: string;
  state: string;
  neighborhood: Neighborhood[];
};

export type Neighborhood = {
  value: string;
  label: string;
};
