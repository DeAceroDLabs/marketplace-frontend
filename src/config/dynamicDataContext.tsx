import React from "react";

const defaultForm = {
  dynamicData: {} as any,
  setDynamicData: (data: any) => {},
};

export const DynamicDataContext = React.createContext(defaultForm);

export const DynamicDataProvider = DynamicDataContext.Provider;

export default DynamicDataContext;
