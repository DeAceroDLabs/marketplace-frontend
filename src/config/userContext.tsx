import React from "react";

const defaultUser = {
  username: "",
  setUser: (username: string) => {},
};

const UserContext = React.createContext(defaultUser);

export const UserProvider = UserContext.Provider;

export default UserContext;
