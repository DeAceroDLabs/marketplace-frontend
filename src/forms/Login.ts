import { Form } from "./form.types";

export const loginForm: Form = {
  form_label: "Login Form",
  fields: [
    {
      name: "login_username",
      label: "Username",
      required: true,
      placeholder: "Username",
      type: "text",
      value: "",
    },
    {
      name: "login_password",
      label: "Password",
      required: true,
      placeholder: "Password",
      type: "text",
      value: "",
    },
  ],
};
