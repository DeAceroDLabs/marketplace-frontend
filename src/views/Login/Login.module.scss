.login-container {
    display: flex;
    flex-direction: row;
}

.form-container {
    width: 80%;
    margin-top: 70px;
}