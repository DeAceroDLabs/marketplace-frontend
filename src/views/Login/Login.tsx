import React, { useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import UserContext from "config/userContext";
import Section from "components/common/Section";
import View from "components/common/View";
import { loginForm } from "forms/Login";

const Login: React.FunctionComponent = () => {
  console.log("login form", loginForm);
  const navigate = useNavigate();
  const { username, setUser } = useContext(UserContext);
  const [user, setUsername] = useState(username);

  const submitUser = () => {
    setUser(user);
    navigate(`../`);
  };

  useEffect(() => {
    if (username) {
      navigate(`../`);
    }
  }, [username, navigate]);

  return (
    <View>
      <Section title="Login">
        <form onSubmit={submitUser}>
          <input
            value={user}
            type="text"
            placeholder="username"
            onChange={(e) => setUsername(e.currentTarget.value)}
          />
          <input type="text" placeholder="password" />
          <button type="submit">Log In</button>
        </form>
      </Section>
    </View>
  );
};

export default Login;
