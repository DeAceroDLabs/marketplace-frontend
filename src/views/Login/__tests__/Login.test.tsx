import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { UserProvider } from "config/userContext";
import { BrowserRouter } from "react-router-dom";
import Login from "../Login";

const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...(jest.requireActual("react-router-dom") as any),
  useNavigate: () => mockedUsedNavigate,
}));

jest.mock("react-hook-form", () => ({
  ...jest.requireActual("react-hook-form"),
  useFormContext: () => ({
    register: () => jest.fn(),
    handleSubmit: () => jest.fn(),
  }),
}));

describe("Login", () => {
  it("loads login without exploding", () => {
    const view = render(
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    );
    expect(view).toMatchSnapshot();
  });

  it("simulates login", async () => {
    render(
      <BrowserRouter>
        <Login />
      </BrowserRouter>
    );

    const username = screen.getByPlaceholderText("Usuario");
    fireEvent.change(username, {
      target: { value: "mock_username" },
    });
    await waitFor(() => {
      expect((username as HTMLInputElement).value).toBe("mock_username");
    });

    const password = screen.getByPlaceholderText("ContraseÃ±a");
    fireEvent.change(password, {
      target: { value: "mock_password" },
    });
    await waitFor(() => {
      expect((password as HTMLInputElement).value).toBe("mock_password");
    });

    const loginButton = screen.getByRole("button");
    fireEvent.click(loginButton);
    await waitFor(() => {
      expect(mockedUsedNavigate).toBeCalled();
    });
  });

  it("simulates user already on context", async () => {
    const mockUSer = {
      username: "mock_user",
      setUser: jest.fn(),
    };

    render(
      <UserProvider value={mockUSer}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </UserProvider>
    );
    await waitFor(() => {
      expect(mockedUsedNavigate).toBeCalled();
    });
  });
});
