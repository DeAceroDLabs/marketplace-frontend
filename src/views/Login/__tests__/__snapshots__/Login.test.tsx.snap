// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login loads login without exploding 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="header"
      >
        <div />
      </div>
      <div
        class="body"
      >
        <div>
          <div
            class="container"
          >
            <div
              class="container"
            >
              <div
                class="title margin-title"
              >
                 
              </div>
              <div
                class="body"
              >
                 
                <div
                  class="form-container"
                >
                  <div
                    class="container"
                  >
                    <form
                      class="form-container"
                    >
                      <div
                        class="container"
                      >
                        <div
                          class="title margin-title"
                        >
                          Ingresa a la plataforma
                           
                        </div>
                        <div
                          class="body"
                        >
                           
                          <p
                            class="form-description"
                          >
                            Ingresa tus datos para acceder a la plataforma
                          </p>
                          <div
                            class="form-container"
                          >
                            <div
                              class="field-container"
                            >
                              <div
                                class="container"
                              >
                                <label>
                                  Usuario
                                </label>
                                <input
                                  class="input"
                                  mock="[object Object]"
                                  placeholder="Usuario"
                                  type="text"
                                  value=""
                                />
                              </div>
                            </div>
                            <div
                              class="field-container"
                            >
                              <div
                                class="container"
                              >
                                <label>
                                  Contrase単a
                                </label>
                                <input
                                  class="input"
                                  mock="[object Object]"
                                  placeholder="Contrase単a"
                                  type="password"
                                  value=""
                                />
                              </div>
                            </div>
                          </div>
                           
                        </div>
                      </div>
                      <button
                        class="button primary"
                      >
                        Ingresar
                      </button>
                    </form>
                  </div>
                </div>
                 
              </div>
            </div>
            <div
              class="vertical"
            >
              <div
                class="content primary right"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="header"
    >
      <div />
    </div>
    <div
      class="body"
    >
      <div>
        <div
          class="container"
        >
          <div
            class="container"
          >
            <div
              class="title margin-title"
            >
               
            </div>
            <div
              class="body"
            >
               
              <div
                class="form-container"
              >
                <div
                  class="container"
                >
                  <form
                    class="form-container"
                  >
                    <div
                      class="container"
                    >
                      <div
                        class="title margin-title"
                      >
                        Ingresa a la plataforma
                         
                      </div>
                      <div
                        class="body"
                      >
                         
                        <p
                          class="form-description"
                        >
                          Ingresa tus datos para acceder a la plataforma
                        </p>
                        <div
                          class="form-container"
                        >
                          <div
                            class="field-container"
                          >
                            <div
                              class="container"
                            >
                              <label>
                                Usuario
                              </label>
                              <input
                                class="input"
                                mock="[object Object]"
                                placeholder="Usuario"
                                type="text"
                                value=""
                              />
                            </div>
                          </div>
                          <div
                            class="field-container"
                          >
                            <div
                              class="container"
                            >
                              <label>
                                Contrase単a
                              </label>
                              <input
                                class="input"
                                mock="[object Object]"
                                placeholder="Contrase単a"
                                type="password"
                                value=""
                              />
                            </div>
                          </div>
                        </div>
                         
                      </div>
                    </div>
                    <button
                      class="button primary"
                    >
                      Ingresar
                    </button>
                  </form>
                </div>
              </div>
               
            </div>
          </div>
          <div
            class="vertical"
          >
            <div
              class="content primary right"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="footer"
    >
      <div />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
